Fine-grained network policies

rylev opened the pull request:
This follows the design outlined in bytecodealliance/wasmtime#7681 for IpNameLookup.

A new trait `IpNameLookup` is introduced that factors out the conversion from a `String` to a list of IP addresses. Another trait, `WasiIpNameLookupView` is used to configure which concrete type that implements `IpNameLookup` will be used in the actual IpNameLookup host implementation. Lastly. the `SystemIpNameLookup` is provided as a default implementation.

Some questions that might guide tweaks:
* Users are now forced to implement `WasiIpNameLookupView` - it would be nice if there was a to default to the built in `SystemIpNameLookup`. If there were default associated types for traits, we could get rid of `WasiIpNameLookupView` and make `ip_name_lookup` a function on `WasiView` that defaults to the the `SystemIpNameLookup`.  But alas that's not available to us.

r? @badeend

rylev committed:
This follows the design outlined in wasmtime#7681 for IpNameLookup.

A new trait `IpNameLookup` is introduced that factors out the conversion
from a `String` to a list of IP addresses. Another trait,
`WasiIpNameLookupView` is used to configure which concrete type that
implements `IpNameLookup` will be used in the actual IpNameLookup host
implementation. Lastly. the `SystemIpNameLookup` is provided as a
default implementation.

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

badeend commented on crates/wasi/src/preview2/ip_name_lookup.rs:
```diff
@@ -5,35 +5,72 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 
-        let host = parse(&name)?;
-
         if !network.allow_ip_name_lookup {
```
With these changes, can `network.allow_ip_name_lookup` now be removed?

rylev replied:
It could be potentially. Right now the default is to not allow ip name lookups, so if we remove this, the default will flip. Additionally, the new way to turn off name lookups would be through a trait implementation (i.e., `impl IpNameLookup for BlockedIpNameLookup` or similar) and not through `WasiCtx`/`WasiCtxBuilder`.

Though it's probably somewhat confusing to implement `WasiIpNameLookupView` and then not have name lookups work because you forgot to flip the bit on `WasiCtx`.

Perhaps we remove the `allow_ip_name_lookup` and provide an additional implementation of `IpNameLookup`: `BlockedIpNameLookup` so that users can easily turn off IP name lookup if they like. Thoughts?

badeend replied:
Hmm. Interesting.
I think the WASI translation layer (`preview2/ip_name_lookup.rs` in this case) should only depend on a single source of truth: `WasiIpNameLookupView`. It should not _also_ go peeking into `network.allow_ip_name_lookup` on its own.

It might be an idea to provide an additional implementation of `IpNameLookup` based on `WasiCtx`, that dynamically checks `allowed_network_uses`. This could supersede the `BlockedIpNameLookup` you suggested

rylev replied:
This does have the downside that the WasiCtx setting can be ignored by users. If their implementation of `IpNameLookup` doesn't check for `allowed_network_uses`, that option will simply be ignored. I guess it seems fine that the user can ignore that option (they're the ones setting it anyway), but it does seem somewhat error prone.

badeend commented on crates/wasi/src/preview2/ip_name_lookup.rs:
```diff
@@ -5,35 +5,72 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
+#[async_trait::async_trait]
+pub trait IpNameLookup {
+    /// Given a name, resolve to a list of IP addresses
+    async fn resolve_addresses(&mut self, name: String) -> std::io::Result<Vec<IpAddr>>;
```
I'm curious; why isn't `resolve_addresses` placed directly  on `WasiIpNameLookupView`? Is there something the extra indirection buys us?

rylev replied:
I'm not sure it gives a ton given that we initialize the type that implements `IpNameLookup` right before we call `resolve_addresses`. The reason I initially went this direction is that it mirrors how the UDP and TCP implementations will work. I can remove it though. Shall I?

badeend replied:
Alright. Was just curious.
No strong preference

rylev committed:
Move the check for allowed_network_uses.ip_name_lookup into the SystemIpNameLookup

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

badeend commented:
All in all, looks good to me 🚀

rylev committed:
Move to a single WasiNetworkView trait

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

rylev commented:
@badeend @alexcrichton 

I've moved the implementation towards having a single `WasiNetworkView` trait where all network based overrides will be implemented. I'm happier with this solution than the previous `IpNameLookupView` version, but I'm still not convinced we've arrived at a good solution. 

The current solution has the following charactertistics:
* Two new methods on `WasiView`: `network_view` and `network_view_mut` which return references to a `dyn WasiNetworkView`
  * This prevents all the methods necessarily being on `WasiView` which might get crowded over time. It also makes it easy for users who want to use an off-the-shelf solution to get implement 2 methods instead of the N methods `WasiNetworkView` will have.
  * The use of dynamic dispatch is new to WasiView but seems appropriate here.
* The various implementors in Wasmtime of `WasiView` gain a new field with `SystemNetwork` (the default implementation of  `WasiNetworkView`)

I would said, we move forward with this for now and we can adjust as we add TCP and UDP support (and of course continue to evolve over time). Thoughts?

alexcrichton commented on crates/wasi/src/preview2/ip_name_lookup.rs:
```diff
@@ -5,35 +5,73 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
+/// A trait for providing a `IpNameLookup` implementation.
+pub trait WasiNetworkView: Send {
+    /// Given a name, resolve to a list of IP addresses
     fn resolve_addresses(
```
This might be best to model as an `async` function rather than returning an `AbortOnDropJoinHandle`?

rylev replied:
This is indeed very annoying... The value returned from `resolve_address` must be `Send + Sync` as it is used in `ResolveAddressStream` which impls `Subscribe` which requires the type implementing it to be `Send + Sync`.

The `#[async_trait]` macro puts a `Send` bound on the return type but not `Sync` and there is no way to instruct it to do so. We could move to using `Box<dyn Future>` but that means we loose the developer experience of `async_trait`. We could potentially move to using `impl Future` but that requires the latest version of Rust, and I'm not sure if that's ok.

alexcrichton replied:
Hm that's true, and yeah I see how there's no great way to manage this right now.

I suppose this can stay for now, but I'm still worried that this will inevitably cause some pain down the road. The best alternative I can think of is to return a `oneshot::Receiver<T>` instead but that's still pretty cumbersome to work with and wouldn't work great here.

rylev replied:
We could take a page out of `WasiHttpView`'s playbook and have the function return a `wasmtime::Result<Resource<ResolveAddressStream>>` just like `WasiHttpView::send_request` returns a `wasmtime::Result<Resource<HostFutureIncomingResponse>>`. I don't really love this either, but at least it's consistent?

badeend replied:
Of all the options mentioned so far, the `Box<dyn Future>` route seems the least worst to me.

The fact that we currently use a quick&dirty implementation using spawn_blocking should be an implementation detail of `SystemNetwork`

alexcrichton replied:
Returning a `Resource<ResolveAddressStream>` would be nice yeah but it seems like that would have all the same design questions around "how do you actually create one of those". 

Although as I say that we could perhaps get around `async trait` by having something like `fn from_future<T: Future + Send + Sync + 'static> (...)` on the `ResolveAddressStream` type. That way you could pass an `async` block which would have bounds generated correctly and additioanlly the handle from `spawn_blocking` could also be passed in? That way we might be able to avoid async traits altogether and rely entirely on that.

rylev replied:
Thinking about this more, returning `Resource<ResolveAddressStream>` would mean that the implementor would have to also implement `WasiCtx` or otherwise have multiple access to the `ResourceTable`. Right now we don't require that from the implementor of `WasiNetworkView`, and I'm not sure we want to.

alexcrichton replied:
`ResolveAddressStream` itself could be returned though leaving the table management bits to the glue?

alexcrichton commented on crates/wasi/src/preview2/ip_name_lookup.rs:
```diff
@@ -5,35 +5,73 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 
-        if !network.allow_ip_name_lookup {
-            return Err(ErrorCode::PermanentResolverFailure.into());
+impl SystemNetwork {
+    /// Create a new `SystemIpNameLookup`
+    pub fn new(ctx: &WasiCtx) -> Self {
+        Self {
+            allowed: ctx.allowed_network_uses.ip_name_lookup,
         }
+    }
+}
```
Could this perhaps manifest as a `impl WasiNetworkView for WasiCtx` to avoid a second type?

alexcrichton replied:
(then it might be neat to have default implementations of `fn network_view` perhaps?)

rylev replied:
Great idea!

rylev committed:
Default impl for network_view methods

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

alexcrichton commented on crates/wasi/src/preview2/ip_name_lookup.rs:
```diff
@@ -5,35 +5,73 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 
-        if !network.allow_ip_name_lookup {
-            return Err(ErrorCode::PermanentResolverFailure.into());
+impl SystemNetwork {
```
Could this be removed in lieu of implementing `WasiNetworkView for WasiCtx` itself? (given that `WasiCtx` is needed to create this structure anyway it may not buy a whole lot)

rylev replied:
If a user wants to have some logic where they conditionally switch between either their own network implementation and `SystemNetwork`, they need some way of constructing the `SystemNetwork`. Without this constructor, they would not be able to do so. We could do the following: 
* make all fields of `SystemNetwork` public
* Make the constructor just take a simple `bool` instead of a `WasiCtx`
* Leave it as is

I think I'd be most in favor of the last two options.

alexcrichton replied:
Hm I'm not sure I understand? Given `impl WasiNetworkView for WasiCtx` then if users are themselves implementing `WasiNetworkView` for their own custom type then conditionally using the default behavior would be calling `self.ctx.the_method_name()` so I'm not sure where the need for having this separate `SystemNetwork` structure arises?

badeend commented:
Having thought about it some more I would like to offer one more suggestion. See code below.
This touches multiple discussions above, so tagging @alexcrichton.

The main difference is that this setup doesn't modify WasiView at all, but instead dispatches through WasiCtx::network

---

> @rylev: This does have the downside that the WasiCtx setting can be ignored by users.

In setup below that can't happen.

---

The strict separation between the "System*" and "Default*" structs below might be overkill for this PR because of the relative trivial "System*" and "Default*" implementations. However, for TCP/UDP sockets the implementation won't be so trivial anymore.

```rs
/// Trait for any WASI-compliant network implementation.
trait Network {
    fn resolve_addresses(&mut self, ) -> ...;

    // fn new_tcp_socket(family) -> ...
    // fn new_udp_socket(family) -> ...
}




/// A WASI-compliant "base" implementation.
/// Without any opinions on how to do filtering, permission checks or whatever.
struct SystemNetwork {}
impl SystemNetwork {
    fn new() -> Self {} // Probably no parameters
}
impl Network for SystemNetwork {
    fn resolve_addresses(&mut self, ) -> {
        // Perform the syscalls without any additional permission checks
    }
}




/// A "good enough" default implementation for Wasmtime users, _with_ permission checks.
struct DefaultNetwork {
    sys: SystemNetwork,
    allowed: bool
}
impl DefaultNetwork {
    /// Whatever parameters we want.
    /// Currently just a simple boolean, but could be extended to any 
    fn new(allowed: bool) -> Self {}
}
impl Network for DefaultNetwork {
    fn resolve_addresses(&mut self, ) -> {
        if self.allowed {
            self.sys.resolve_addresses()
        } else {
            Err(PermissionDenied)
        }
    }
}



pub struct WasiCtx {
    pub(crate) random: Box<dyn RngCore + Send + Sync>,
    pub(crate) insecure_random: Box<dyn RngCore + Send + Sync>,
    pub(crate) insecure_random_seed: u128,
    pub(crate) wall_clock: Box<dyn HostWallClock + Send + Sync>,
    pub(crate) monotonic_clock: Box<dyn HostMonotonicClock + Send + Sync>,
    pub(crate) env: Vec<(String, String)>,
    pub(crate) args: Vec<String>,
    pub(crate) preopens: Vec<(Dir, String)>,
    pub(crate) stdin: Box<dyn StdinStream>,
    pub(crate) stdout: Box<dyn StdoutStream>,
    pub(crate) stderr: Box<dyn StdoutStream>,

    // Remove `socket_addr_check`
    // Remove `allowed_network_uses`
    
    // Add:
    pub(crate) network: Box<dyn Network>,
}





pub struct WasiCtxBuilder {
    // (...)

    network: Box<dyn Network>,
}
impl WasiCtxBuilder {
    pub fn new() -> Self {
        Self {
            // (...)
            network: DefaultNetwork::new(/*allowed: */false),
        }
    }

    pub fn allow_network(&mut self, enable: bool) -> &mut Self {
        self.network = DefaultNetwork::new(enable);
        self
    }

    pub fn custom_network(&mut self, net: Box<dyn Network>) -> &mut Self {
        self.network = net;
        self
    }
}
```

alexcrichton commented:
@badeend your sketch makes sense yeah, but my thoughts at https://github.com/bytecodealliance/wasmtime/issues/7694#issuecomment-1858351478 showcase how something like that breaks down unless all callbacks about configuration go through `trait Network`, so for example we'd also need to route IP address checks through that trait as well. (not a problem of course, just pointing out that once things are behind a trait everything needs to be behind that trait to share state between callbacks).

Otherwise though I think it'd be reasonable to configure a trait object on `WasiCtx` vs returning a `&mut dyn Trait` from `WasiView` (although it having a default impl returning `self.ctx_mut()` I think is pretty nifty and makes it almost hidden.

rylev committed:
A trait object `Network`

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

rylev commented:
I like @badeend's suggestion. I initially shared the same concern that @alexcrichton pointed out, but I think it should be relatively easy for users to implement the `Network` trait by delegating all the parts they don't care about to `SystemNetwork` or `DefaultNetwork` and only changing what they do care about. The only issue I can see with this is that if the `Network` trait grows large (which I think we anticipate it will), then there will be a large amount of boiler plate of delegating `Network` methods to the `SystemNetwork` or `DefaultNetwork`. This could be really annoying.

rylev commented:
@alexcrichton @badeend I've pushed a new change that moves more towards @badeend's design as well as changing the return type of `resolve_addresses` to be a `ResolveAddressStream`. 

I think the only way for us to be sure about this design is to keep going and implement UDP and TCP on top of it. With that in mind, I think we should merge this PR and continue the design exploration in a follow up. Thoughts?

badeend commented:
>  something like that breaks down unless all callbacks about configuration go through trait Network, so for example we'd also need to route IP address checks through that trait as well.

Correct. In my mind that was exactly the goal of this (and upcoming) PRs. `socket_addr_check` & `allowed_network_uses` would most likely be moved to `DefaultNetwork`

> I think the only way for us to be sure about this design is to keep going and implement UDP and TCP on top of it.

Agree.

> With that in mind, I think we should merge this PR and continue the design exploration in a follow up. Thoughts?

I'm not familiar with wasmtime's release process, but if a release gets cut partway through these changes it could be weird for consumers when e.g. IP lookup & TCP use the Network trait but UDP still uses an old & completely different customization mechanism. I'll leave this up for @alexcrichton

alexcrichton commented on crates/wasi/src/preview2/ip_name_lookup.rs:
```diff
@@ -2,39 +2,44 @@ use crate::preview2::bindings::sockets::ip_name_lookup::{Host, HostResolveAddres
 
 #[async_trait::async_trait]
-impl<T: WasiView> Host for T {
+impl<T: WasiView + Sized> Host for T {
```
I think the `+ Sized` may not be necessary here?

alexcrichton reviewed (commented):
Looks reasonable to me!

Unless this is particularly urgent though I would personally agree with @badeend in that I'd prefer to see a more complete picture before landing if possible. If this is needed to unblock work I think it's ok to land, but if you're ok working with a fork for now though I think it'd be best to sketch out the other checks you're thinking of implementing to be able to see them altogether at once.

rylev commented:
I'm fine with not landing this on main. I just think we need to decide which fork/branch is the integration point so that if @badeend and I continue the work, we can bring that all together into one unified branch that will ultimately be merged into main.

badeend committed:
Reorder code blocks

badeend committed:
Move tcp_create_socket impl

badeend committed:
Move preview2/tcp.rs into preview2/host/tcp.rs

badeend committed:
Rename types to include *Wrapper suffix

badeend committed:
Migrate to a AsyncRead+AsyncWrite based implementation

badeend committed:
Define our own SocketAddrFamily so that we don't have a public dependency on cap_net_ext

badeend committed:
Use std::io::Result for the validation utils

badeend committed:
Introduce SystemTcpSocket type with bare-bones implementation

badeend committed:
Move the low hanging fruit out of the WASI wrapper and into SystemTcpSocket.

Also updated some overly strict test cases.

badeend committed:
Move implementation of listen & listen_backlog_size into SystemTcpSocket

badeend committed:
Explicitly specify which states are awaitable

badeend committed:
Extract out the async polling mechanism behind `accept`. Similar to tokio::net::TcpListener::poll_accept

badeend committed:
Separate IP name lookup implementation out from the WASI translation layer

badeend committed:
Introduce DynFuture

badeend committed:
Refactor IP name lookup to use DynFuture, taking advantage of the non-allocating variant wherever possible.

badeend committed:
Move `connect` into SystemTcpSocket

badeend committed:
Move `bind` into SystemTcpSocket

badeend committed:
Inline TCP-specific utilities

badeend committed:
Return the streams from connect&accept

badeend committed:
Refactor `DynFuture` away into two types:
- BoxSyncFuture: for public facing signatures. This is just a regular `Pin<Box<dyn Future ...>>`
- Preview2Future: utility type for use within the WASI wrapper code

badeend committed:
Introduce TcpSocket trait

badeend committed:
Refactor:
- Remove AllowedNetworkUses in favor of just socket_addr_checks
- Let all permission checks go through self.ctx(), instead of NetworkHandle

badeend committed:
Implement DefaultTcpSocket

badeend committed:
Fix error in MacOS-only code

badeend committed:
Resolve ambiguity between SocketAddrFamily and SocketAddressFamily

badeend committed:
Update comments

badeend committed:
Document why these public methods exist

badeend committed:
Explicitly call it `deny`

badeend committed:
Update crates/wasi/src/preview2/host/network.rs

Co-authored-by: Ryan Levick <me@ryanlevick.com>

badeend committed:
Update crates/wasi/src/preview2/host/network.rs

Co-authored-by: Ryan Levick <me@ryanlevick.com>

badeend committed:
Merge branch 'tcp-override' of https://github.com/badeend/wasmtime into tcp-override

badeend committed:
Give all WASI wrappers the same suffix

badeend committed:
Eliminate fall-throughs from `resolve_next_address`

rylev committed:
Merge pull request #2 from badeend/tcp-override

TCP overrides

badeend committed:
Merge branch 'main' of https://github.com/bytecodealliance/wasmtime into tcp-override

badeend committed:
After 9124e03b1ab42e2270996344e155049074848e50  there's no distinction between SocketProtocolMode and SocketAddrFamily anymore

rylev committed:
Merge pull request #3 from badeend/tcp-override

Sync with upstream

rylev committed:
Move everything over from UDP but streams

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

rylev committed:
Get it all compiling

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

rylev committed:
Allow overriding streams

Signed-off-by: Ryan Levick <ryan.levick@fermyon.com>

rylev committed:
Merge pull request #4 from rylev/udp-override

UDP override

badeend commented:
@alexcrichton One of the (many..) changes in this PR is that the actual socket code has been separated out from the component trait implementation to its own `SystemTcpSocket`/`SystemUdpSocket`. To reduce the size of this PR, do you mind if I split that refactor off from this PR and submit a new one with just those changes? These changes will only be internal reshuffling of code, without affecting the public interface.

alexcrichton commented:
Sounds good to me! And yeah refactorings are always good to land at any time
